### PR TITLE
gh-105323: Update readline module to detect apple editline variant

### PR DESCRIPTION
**Pull request by corona10**
----
:books: Documentation preview :books:: https://cpython-previews--108665.org.readthedocs.build/

* Issue: gh-105323

### CODE REVIEW

**Comment by ned-deily** [comment]
Without having time to look closely at this now. I feel pretty strongly against adding a new configuration option for what has always worked by default before. There should be a better way.

**Comment by corona10** [comment]
> Without having time to look closely at this now. I feel pretty strongly against adding a new configuration option for what has always worked by default before. There should be a better way.

Yeah let's discuss it after your travel :) Enjoy!

**Comment by corona10** [comment]
>  I feel pretty strongly against adding a new configuration option for what has always worked by default before.

The different signature has existed since macOS 10.5. 
We didn't notice for a long time, and it was never been a issue because the compiler didn't detected..
but I think that this is a correct way to handle Apple issue.

**Comment by erlend-aasland** [comment]
I agree with @ned-deily. Before proposing new PRs, let's instead lay out the three scenarios in detail and _then_ see how we can improve configure and readline.c preprocessor guards :)

**Comment by corona10** [comment]
@erlend-aasland @ned-deily 

| Backend        	| Real verison 	| Macro Version 	| Intall Source   	| rl_startup_hook                            	| rl_pre_input_hook                          	|
|----------------	|--------------	|---------------	|-----------------	|--------------------------------------------	|--------------------------------------------	|
| readline       	| 8.2.1        	| 0x0802        	| brew / MacPorts 	| typedef int rl_hook_func_t (void);         	| typedef int rl_hook_func_t (void);         	|
| editline       	| 20221030-3.1 	| 0x0402        	| brew / MacPorts 	| rl_hook_func_t(void);                      	| rl_hook_func_t(void);                      	|
| Apple editline 	| n/a          	| 0x0402        	| XCode SDK       	| typedef int 	  Function(const char *, int); 	| typedef int 	  Function(const char *, int); 	|

First of all, I love seamless approach, but here is the table for why 
we should handle Apple editline as an independent backend.
Apple editline only shares the same header name, and it uses fully different signature and implementation. 

There is no way to determine between editline and Apple editline except using `grep` (I don't prefer this option it's too unstable) because they are using same macro versions.

Suppose Apple editline is viewed as an independent implementation. In that case, I think it is better in terms of consistency to treat it separately as an Apple option as if the Edlitline option was added.

**Comment by erlend-aasland** [comment]
Hm, I wonder if there are really four flavours we need to take into account:

- GNU readline (readline/readline.h)
- libedit w/readline wrapper (editline/readline.h)
- libedit (editline/editline.h)
- Apple libedit (editline/readline.h _and_ readline/readline.h, both identical)

**Comment by corona10** [comment]
> libedit w/readline wrapper (editline/readline.h)
> libedit (editline/editline.h)

When I refer the [libedit hompage](https://www.thrysoee.dk/editline/), `editline/readline.h` only exists.
I am not sure where I can refer the `libedit (editline/editline.h)` as official case.

**Comment by erlend-aasland** [comment]
> > libedit w/readline wrapper (editline/readline.h)
> > libedit (editline/editline.h)
> 
> When I refer the [libedit hompage](https://www.thrysoee.dk/editline/), `editline/readline.h` only exists. I am not sure where I can refer the `libedit (editline/editline.h)` as official case.

Minix Editline (https://troglobit.com/projects/editline/) provides editline/editline.h.

**Comment by corona10** [comment]
> Minix Editline (https://troglobit.com/projects/editline/) provides editline/editline.h.

Thanks @erlend-aasland, I updated the table :)

**Comment by ronaldoussoren** [comment]
> Suppose Apple editline is viewed as an independent implementation. In that case, I think it is better in terms of consistency to treat it separately as an Apple option as if the edlitline option was added.

Apple's lib edit isn't really an independent implementation, but is a very old copy of NetBSD libedit. [This old revision](http://cvsweb.netbsd.org/bsdweb.cgi/src/lib/libedit/readline/readline.h?rev=1.33.2.1&content-type=text/x-cvsweb-markup) in the NetBSD CVS(!) server has similar definitions.

That said, this is just nitpicking on your statement. For handling the differences you might as well treat this as the "apple" implementation because I expect nobody else uses the old code base. 

Do you know if the newer libedit is consistently better than Apple's copy? If it is we could consider to just ship a current version with our installers, but that's for a different issue.

**Comment by corona10** [comment]
> Apple's lib edit isn't really an independent implementation, but is a very old copy of NetBSD libedit. [This old revision](http://cvsweb.netbsd.org/bsdweb.cgi/src/lib/libedit/readline/readline.h?rev=1.33.2.1&content-type=text/x-cvsweb-markup) in the NetBSD CVS(!) server has similar definitions.

Oh I thought that because they manage independent repo but it looks not?
https://github.com/apple-oss-distributions/libedit/

**Comment by corona10** [comment]
> Do you know if the newer libedit is consistently better than Apple's copy? If it is we could consider to just ship a current version with our installers, but that's for a different issue.

Not sure :(

**Comment by corona10** [comment]
>  If it is we could consider to just ship a current version with our installers, but that's for a different issue.

Yeah, we already did for libexpat, so we can consider the vendoring the library if it is worth doing it.
But it will require maintenance costs as same as libexpat, so I agree that it will be a completely different issue.

**Comment by corona10** [comment]
@ronaldoussoren @ned-deily @erlend-aasland 

Hi team :)
I would like to merge this PR if there is no better solution.
This change is revertable everytime and you can still use the legacy configuration(just `./configure`) if needed.

WDYT?

**Comment by ned-deily** [comment]
I haven't finished traveling yet so I haven't reviewed it yet. Please wait a few more days.

**Comment by ned-deily** [comment]
See also new issue #109191.

**Comment by erlend-aasland** [comment]
I agree with Ned. There is no need to rush a partial solution.

**Comment by corona10** [comment]
> I agree with Ned. There is no need to rush a partial solution.

Well if we find a better solution it will be choosen but if not I would like to say this is best effort with non complex solution.

The solution looks partial because apple distributed the library partial changed version.

**Comment by corona10** [comment]
We can depends on specific API is existed or not but do we have to track all Apple changes even if they fork from 2007?

As I summerized above they even use same version sementics.

**Comment by corona10** [comment]
To Ned, I am not rushing. I just pinged you since this PR was opened 2 weeks ago.
I can wait :) And if there is better solution I can close the PR.

**Comment by ronaldoussoren** [comment]
The patch looks sane, although I haven't tested it myself.  

That said, it should be fairly easy to detect the ``apple`` variant in the configure phase if we don't mind shipping an ``AC_CHECK_TYPEDEF`` (http://doc.gnu-darwin.org/ac-archive/guidod/ac_check_typedef.html) by checking for ``Function`` typedef. That way the invocation of configure doesn't have to change.  I'm not currently set up to be able to write a new check in ``configure.ac`` though.

**Comment by erlend-aasland** [comment]
> That said, it should be fairly easy to detect the `apple` variant in the configure phase if we don't mind shipping an `AC_CHECK_TYPEDEF` (http://doc.gnu-darwin.org/ac-archive/guidod/ac_check_typedef.html) by checking for `Function` typedef. That way the invocation of configure doesn't have to change. I'm not currently set up to be able to write a new check in `configure.ac` though.

Sounds like a plan that aligns with Ned's wish regarding things working automagically.

File: Modules/readline.c
Comment (erlend-aasland) [inline Modules/readline.c]: Would it not be easier to use `Py_UNUSED` on the args instead of the (void) trick?

**Comment by corona10** [comment]
I am testing `AC_CHECK_TYPEDEF`, but we may need to find more modern macro :)
(I am finding)
```
configure.ac:5835: warning: The macro `AC_CHECK_TYPEDEF' is obsolete.
````

**Comment by corona10** [comment]
Looks like the following macro will work: https://www.gnu.org/software/autoconf-archive/ax_check_typedef.html
```
AX_CHECK_TYPEDEF(Function, editline/readline.h, [,AC_DEFINE([WITH_APPLE_EDITLINE]),])
```
I am still testing

**Comment by corona10** [comment]
Okay using `AX_CHECK_TYPEDEF ` can provide seamless solution
@ned-deily @ronaldoussoren @erlend-aasland PTAL

File: configure.ac
Comment (corona10) [inline configure.ac]: I am not sure that we have to support for `AS_VAR_IF([with_readline], [edit], [...])` case too

**Comment by corona10** [comment]
Oops, I need more testing.

File: configure.ac
Comment (corona10) [inline configure.ac]: @erlend-aasland  cc @ronaldoussoren 

I am not sure why AC_DEFINE([WITH_APPLE_EDITLINE]) is not generated in configure file.
Checking itself looks okay but the problem is if the condition is true it doesn't do as my expectation.

````
checking whether right shift extends the sign bit... yes
checking for getc_unlocked() and friends... yes
checking for Function in readline/readline.h... yes
````

**Comment by corona10** [comment]
@ronaldoussoren @ned-deily @erlend-aasland 

Finally, I updated the PR as the working solution. I checked both cases.
PTAL

FYI, I found some comments about AX_CHECK_TYPEDEF / AC_CHECK_TYPEDEF are broken, so they recommend to use 
AC_CHECK_TYPE

File: configure.ac
Comment (corona10) [inline configure.ac]: @erlend-aasland 

Is it correct to write the header in this way?
IIRC you recommend another way to write a header in configure.ac.

**Comment by corona10** [comment]
@ronaldoussoren @ned-deily @erlend-aasland 

gentle ping!

File: Modules/readline.c
Comment (erlend-aasland) [inline Modules/readline.c]: `__APPLE__` is implied if `WITH_APPLE_EDITLINE` is defined, so this should suffice:

```suggestion
#elif defined(WITH_APPLE_EDITLINE)
```

Comment (erlend-aasland) [reply Modules/readline.c]: Also, keeping the opening brace out of the ifdef should result in a cleaner diff?

File: configure.ac
Comment (erlend-aasland) [inline configure.ac]:
```suggestion
# gh-105323: Need to handle the macOS editline as an alias of readline.
```

Let's keep the comment; it makes it easier to debug the generated file.

File: configure.ac
Comment (erlend-aasland) [reply configure.ac]:
```suggestion
  [Darwin/*], [AC_CHECK_TYPE([Function],
                             [AC_DEFINE([WITH_APPLE_EDITLINE])],
                             [],
                             [@%:@include <readline/readline.h>])],
```

File: Doc/using/configure.rst
Comment (erlend-aasland) [inline Doc/using/configure.rst]:
```suggestion
   Designate a backend library for the :mod:`readline` module.
```

**Review by erlend-aasland — APPROVED** [review]
Thanks, Dong-hee; this looks good to me! (And thanks for looking into these readline issues!) AFAICS, this looks like a solution that should please all involved core devs. Please wait for a thumbs-up from Ronald and the other Ned before merging 🙂

File: Modules/readline.c
Comment (corona10) [inline Modules/readline.c]:
```suggestion
on_startup_hook(const char *Py_UNUSED(text), int Py_UNUSED(state))
```

**Review by ronaldoussoren — APPROVED** [review]
LGTM

**Pull request merged**
